File: src/main/java/com/leetcode/top/interview/easy/Sqrt.java
package com.leetcode.top.interview.easy;

import org.junit.Assert;
import org.junit.Test;

/**
 * @author wezhyn
 * @since 04.04.2020
 */
public class Sqrt {

    @Test
    public void test() {
        Assert.assertEquals(2,mySqrt(8));
    }
    public int mySqrt(int x) {
        return (int) Math.sqrt(x);
    }
}

File: src/main/java/com/leetcode/bs/Sqrt.kt
package com.leetcode.bs

/**
 *Implement int sqrt(int x).

Compute and return the square root of x, where x is guaranteed to be a non-negative integer.

Since the return type is an integer, the decimal digits are truncated and only the integer part of the result is returned.
 * 使用值二分的思想
 * @author wezhyn
 * @since 08.19.2020
 *
 */
class Sqrt {
    fun mySqrt(x: Int): Int {
        var l = 1
        var r = x
        while (l <= r) {
            val middle = l + (r - l) / 2
            val dm = middle * middle.toLong()
            if (dm <= 0) {
                r = middle - 1
                continue
            }
            if (dm > x) {
                r = middle - 1
            } else if (dm == x.toLong()) {
                return middle
            } else {
                l = middle + 1
            }
        }
        return r
    }

}